test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_work_dim.cpp

```diff
@@ -0,0 +1,223 @@
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <extensionHelpers.h>
+#include "mutable_command_basic.h"
+
+#include <array>
+#include <cstring>
+
+#include <CL/cl_ext.h>
+
+// mutable dispatch tests setting `work_dim` to the original 3D value
+// behaves as expected.
+
+struct MutableDispatchWorkDim : public InfoMutableCommandBufferTest
+{
+    using InfoMutableCommandBufferTest::InfoMutableCommandBufferTest;
+
+    MutableDispatchWorkDim(cl_device_id device, cl_context context,
+                           cl_command_queue queue)
+        : InfoMutableCommandBufferTest(device, context, queue)
+    {}
+
+    cl_int SetUp(int elements) override
+    {
+        result_data.resize(update_elements);
+        return InfoMutableCommandBufferTest::SetUp(elements);
+    }
+
+    bool Skip() override
+    {
+        cl_mutable_dispatch_fields_khr mutable_capabilities;
+
+        bool mutable_support =
+            !clGetDeviceInfo(
+                device, CL_DEVICE_MUTABLE_DISPATCH_CAPABILITIES_KHR,
+                sizeof(mutable_capabilities), &mutable_capabilities, nullptr)
+            && (mutable_capabilities & CL_MUTABLE_DISPATCH_GLOBAL_SIZE_KHR);
+
+        return !mutable_support || InfoMutableCommandBufferTest::Skip();
+    }
+
+    bool Verify(cl_mem buffer, cl_uint gid_elements)
+    {
+        std::memset(result_data.data(), 0, alloc_size);
+        cl_int error =
+            clEnqueueReadBuffer(queue, buffer, CL_TRUE, 0, alloc_size,
+                                result_data.data(), 0, nullptr, nullptr);
+        test_error(error, "clEnqueueReadBuffer failed");
+
+        for (size_t i = 0; i < gid_elements; i++)
+        {
+            if (result_data[i] != gid_elements)
+            {
+                log_error("Data failed to verify at index %zu. "
+                          "Expected %u, result was %u\n",
+                          i, gid_elements, result_data[i]);
+                return false;
+            }
+        }
+        return true;
+    }
+
+    cl_int Run() override
+    {
+        const char *global_size_kernel =
+            R"(
+                __kernel void three_dim(__global uint *dst0,
+                                        __global uint *dst1,
+                                        __global uint *dst2)
+            {
+                size_t gid0 = get_global_id(0);
+                dst0[gid0] = get_global_size(0);
+
+                size_t gid1 = get_global_id(1);
+                dst1[gid1] = get_global_size(1);
+
+                size_t gid2 = get_global_id(2);
+                dst2[gid2] = get_global_size(2);
+            })";
```

Comment:
This probably won't matter in practice, but strictly speaking I think there is a data race here, since multiple work-items will share the same global ID in each dimension for a 3D dispatch.

Options:
1. Convince oursleves that this doesn't matter in practice and do nothing.
2. Switch to a 1D dispatch instead, so each global ID is unique.
3. Use a linear global ID instead.